src/linker/Linker/MessageOrigin.cs

```diff
@@ -20,29 +23,45 @@ public MessageOrigin (string fileName, int sourceLine = 0, int sourceColumn = 0)
 			FileName = fileName;
 			SourceLine = sourceLine;
 			SourceColumn = sourceColumn;
+			MdTokenProvider = null;
 		}
 
-		public static MessageOrigin? TryGetOrigin (IMemberDefinition sourceMethod, int ilOffset)
+		public MessageOrigin (IMetadataTokenProvider mdTokenProvider, int sourceLine = 0, int sourceColumn = 0)
 		{
-			if (sourceMethod is MethodDefinition methodDef) {
-				if (!methodDef.DebugInformation.HasSequencePoints)
-					return null;
+			FileName = null;
+			MdTokenProvider = mdTokenProvider;
+			SourceLine = sourceLine;
+			SourceColumn = sourceColumn;
+		}
+
+		private MessageOrigin (string fileName, IMetadataTokenProvider mdTokenProvider, int sourceLine = 0, int sourceColumn = 0)
+		{
+			FileName = fileName;
+			MdTokenProvider = mdTokenProvider;
+			SourceLine = sourceLine;
+			SourceColumn = sourceColumn;
+		}
 
-				SequencePoint correspondingSequencePoint = methodDef.DebugInformation.SequencePoints
+		public static MessageOrigin TryGetOrigin (MethodDefinition sourceMethod, int ilOffset)
```

Comment:
This is quite expensive and you call it before warning is rejected. Could this be rewritten to be done when the FileName is actually needed?

Reply:
Yes, we can do this and we should. I've opened an issue for this: https://github.com/mono/linker/issues/1221